network/endpoint_linux.go

```diff
@@ -53,205 +54,200 @@ func (nw *network) newEndpointImpl(
+			} else if nw.Mode != opModeTransparent {
+				logger.Info("Bridge client")
+				epClient = NewLinuxBridgeEndpointClient(nw.extIf, hostIfName, contIfName, nw.Mode, nl, plc)
+			} else if epInfo.NICType == cns.DelegatedVMNIC {
+				logger.Info("Secondary client")
+				epClient = NewSecondaryEndpointClient(nl, plc, ep)
 			} else {
-				logger.Info("OVS client")
-				if _, ok := epInfo.Data[SnatBridgeIPKey]; ok {
-					nw.SnatBridgeIP = epInfo.Data[SnatBridgeIPKey].(string)
-				}
-
-				epClient = NewOVSEndpointClient(
-					nw,
-					epInfo,
-					hostIfName,
-					contIfName,
-					vlanid,
-					localIP,
-					nl,
-					ovsctl.NewOvsctl(),
-					plc)
+				logger.Info("Transparent client")
+				epClient = NewTransparentEndpointClient(nw.extIf, hostIfName, contIfName, nw.Mode, nl, plc)
 			}
-		} else if nw.Mode != opModeTransparent {
-			logger.Info("Bridge client")
-			epClient = NewLinuxBridgeEndpointClient(nw.extIf, hostIfName, contIfName, nw.Mode, nl, plc)
-		} else {
-			logger.Info("Transparent client")
-			epClient = NewTransparentEndpointClient(nw.extIf, hostIfName, contIfName, nw.Mode, nl, plc)
 		}
-	}
 
-	// Cleanup on failure.
-	defer func() {
-		if err != nil {
-			logger.Error("CNI error. Delete Endpoint and rules that are created", zap.Error(err), zap.String("contIfName", contIfName))
-			endpt := &endpoint{
-				Id:                       epInfo.Id,
-				IfName:                   contIfName,
-				HostIfName:               hostIfName,
-				LocalIP:                  localIP,
-				IPAddresses:              epInfo.IPAddresses,
-				DNS:                      epInfo.DNS,
-				VlanID:                   vlanid,
-				EnableSnatOnHost:         epInfo.EnableSnatOnHost,
-				EnableMultitenancy:       epInfo.EnableMultiTenancy,
-				AllowInboundFromHostToNC: epInfo.AllowInboundFromHostToNC,
-				AllowInboundFromNCToHost: epInfo.AllowInboundFromNCToHost,
+		//nolint:gocritic
+		defer func(client EndpointClient, contIfName string) {
+			// Cleanup on failure.
+			if err != nil {
+				logger.Error("CNI error. Delete Endpoint and rules that are created", zap.Error(err), zap.String("contIfName", contIfName))
+				if containerIf != nil {
+					client.DeleteEndpointRules(ep)
+				}
+				// set deleteHostVeth to true to cleanup host veth interface if created
+				//nolint:errcheck // ignore error
+				client.DeleteEndpoints(ep)
 			}
+		}(epClient, contIfName)
 
-			if containerIf != nil {
-				endpt.MacAddress = containerIf.HardwareAddr
-				epClient.DeleteEndpointRules(endpt)
+		//nolint:wrapcheck // ignore wrap check
+		err = func() error {
+			if epErr := epClient.AddEndpoints(epInfo); epErr != nil {
+				return epErr
 			}
 
-			if nw.extIf != nil {
-				endpt.Gateways = []net.IP{nw.extIf.IPv4Gateway}
+			if epInfo.NICType == cns.InfraNIC {
+				var epErr error
+				containerIf, epErr = netioCli.GetNetworkInterfaceByName(contIfName)
+				if epErr != nil {
+					return epErr
+				}
+				ep.MacAddress = containerIf.HardwareAddr
```

Comment:
why can't this be moved under appropriate client code?

Reply:
this code should be part of AddEndpoints api

Reply:
we still have to set the endpoint MacAddress based on the results

ep.MacAddress = containerIf.HardwareAddr

and the defer also checks containerIf before calling client.DeleteEndpointRules

I think we can get container mac from the ep client but still need to check for default address type and update all the clients to return the container mac or add a func to retrieve it

ep.MacAddress = epClient.containerMac

In this case can we check if mac is empty to call epClient.DeleteEndpointRules?